File: src/com/CopperPenguin96/Redstone/System/Types/OptValue.java
package com.CopperPenguin96.Redstone.System.Types;

public class OptValue<T> {

	private boolean _enabled;
	public boolean isEnabled() {
		return _enabled;
	}
	
	public void setEnabled(boolean enabled) {
		_enabled = enabled;
	}
	
	public void setEnabled() {
		_enabled = true;
	}
	
	private T _value;
	
	public void setValue(T val) {
		_value = val;
	}
	
	public T getValue() {
		return _value;
	}
	
	/**
	 * Assumes nothing
	 * @param enabled
	 * @param value
	 */
	public OptValue(boolean enabled, T value) {
		setEnabled(enabled);
		setValue(value);
	}
	
	/**
	 * Assumes value is preset.
	 * @param value
	 */
	public OptValue(T value) {
		setEnabled();
		setValue(value);
	}
	
	/**
	 * Assumes value is not present.
	 */
	public OptValue() {
		setEnabled(false);
	}
}

File: RedstoneCore/Types/OptValue.cs
namespace Redstone.Core.Types
{
    public class OptValue<T>
    {
        public bool Enabled { get; set; }

        private T? _value;
        public T Value
        {
            get
            {
                if (!Enabled) return default!;
                return _value!;
            }
            set
            {
                _value = value;
            }
        }

        public OptValue(bool enabled = false, T? value = default)
        {
            Enabled = enabled;
            Value = value!;
        }

        public OptValue(T value) : this(true, value) { }
    }
}
